### PR TITLE
Adding UserDTO,HunterDTO,UserAuthenticateDTO,refactoring structure of…

### PR DESCRIPTION
**Pull request by erikmacej**
… api module

### CODE REVIEW

File: api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package cz.fi.muni.pa165.dto;
+
+/**
+ * Created by Erik Macej 433744 , on 19.11.16.
+ *
+ * @author Erik Macej 433744
+ */
+public class UserAuthenticateDTO {
+
+    private String passwordHash;
```
Comment (atiratree) [inline api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java]: The client should not implement hashing. It is less secure and problematic for future refactoring.
It should be done on the server side; i.e. exchange just passwords instead.

File: api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    private String nick;
```
Comment (atiratree) [inline api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java]: I think emails should be used for authentication; not nicks.

File: api/src/main/java/cz/fi/muni/pa165/dto/UserDTO.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    @JsonView(View.Summary.class)
+    private String surname;
+
+    @JsonView(View.Summary.class)
+    private String nick;
+
+    @JsonView(View.Summary.class)
+    private String email;
+
+    private String passwordHash;
```
Comment (atiratree) [inline api/src/main/java/cz/fi/muni/pa165/dto/UserDTO.java]: what is the reason for sending  hashes when querying users?

File: api/src/main/java/cz/fi/muni/pa165/dto/HunterDTO.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Created by Erik Macej 433744 , on 19.11.16.
+ *
+ * @author Erik Macej 433744
+ */
+public class HunterDTO extends UserDTO {
```
Comment (atiratree) [inline api/src/main/java/cz/fi/muni/pa165/dto/HunterDTO.java:16]: I am not sure if we want to register users through HunterDTO or to create DTO especially for registering...

Comment (atiratree) [reply api/src/main/java/cz/fi/muni/pa165/dto/HunterDTO.java:16]: We could leave it like this I guess, especially if we will be using @JsonView for that

File: api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java (hunk 2 above)
Comment (atiratree) [reply api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java]: Or at least both: emails or nicks

File: api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java (hunk 1 above)
Comment (erikmacej) [reply api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java]: aha chapem to bola len chyba z nepozornosti

File: api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java (hunk 2 above)
Comment (erikmacej) [reply api/src/main/java/cz/fi/muni/pa165/dto/UserAuthenticateDTO.java]: tiez chyba z nepozornosti

File: api/src/main/java/cz/fi/muni/pa165/dto/UserDTO.java (hunk 3 above)
Comment (erikmacej) [reply api/src/main/java/cz/fi/muni/pa165/dto/UserDTO.java]: Hashe sa v responsoch posielat nebudu, na to je ta JsonView ale pre pracu s UserDto to potrebujes

Comment (atiratree) [reply api/src/main/java/cz/fi/muni/pa165/dto/UserDTO.java]: proč to tam potřebuješ když se to nebude posílat?

Comment (erikmacej) [reply api/src/main/java/cz/fi/muni/pa165/dto/UserDTO.java]: Tak asi naozaj ho tam nebude treba. Predtym som si myslel ze ho ptrebujem v HunterFacade ale asi fakt nie. Tak neviem ci ten Json view tam ma este nejaky zmysel. Mozno iba pre ten ze sa tam zbytocne neposielal ten List<Visit> ale inak neviem ci to ma dajaky zmysel.

Comment (erikmacej) [reply api/src/main/java/cz/fi/muni/pa165/dto/UserDTO.java]: Ale na druhej strane by som ho tam nechal, aspon v tomto milestone. Ked budeme robit milestone 3 uvidime ci to  nejako vyuzijeme ak nie tak sa zmaze

Comment (atiratree) [reply api/src/main/java/cz/fi/muni/pa165/dto/UserDTO.java]: J kvůli tomu listu se to vyplatí. 
Opravdu mě nenapadá jediný důvod proč to tam má být. Přístup by měl být spíš opačný, tj. přidat to tam, až to budeš potřebovat.

Comment (erikmacej) [reply api/src/main/java/cz/fi/muni/pa165/dto/UserDTO.java]: Dobre tak passwordHash damo s DTO prec a json view tam necham

File: api/src/main/java/cz/fi/muni/pa165/dto/HunterDTO.java (hunk 4 above)
Comment (erikmacej) [reply api/src/main/java/cz/fi/muni/pa165/dto/HunterDTO.java:16]: Mam to menit alebo to mozem nechat tak ?

File: api/src/main/java/cz/fi/muni/pa165/facade/HunterFacade.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ * @author Erik Macej 433744
+ */
+public interface HunterFacade {
+
+    /**
+     * Register given hunter with his password
+     * @param hunter - hunter to register
+     * @param unencryptedPassword - hunter's password
+     */
+    void registerHunter(HunterDTO hunter, String unencryptedPassword) throws InvalidKeySpecException,
```
Comment (atiratree) [inline api/src/main/java/cz/fi/muni/pa165/facade/HunterFacade.java]: better to unify into one exception, i.e. AuthenticationException

File: api/src/main/java/cz/fi/muni/pa165/facade/HunterFacade.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            NoSuchAlgorithmException;
+
+    /**
+     * Authenticates Hunter
+     * @return - true only if hashed password is equal with hunter passwordHash
+     */
+    boolean authenticate(HunterDTO hunter, String password) throws InvalidKeySpecException,
```
Comment (atiratree) [inline api/src/main/java/cz/fi/muni/pa165/facade/HunterFacade.java]: same

File: api/src/main/java/cz/fi/muni/pa165/facade/HunterFacade.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+     * Finds all Hunters from Database
+     *
+     * @return list of Hunters
+     */
+    Collection<HunterDTO> findAll();
+
+    /**
+     * Changes user password
+     */
+    void changePassword(HunterDTO hunter,String oldUnencryptedPassword, String newUnencryptedPassword)
```
Comment (atiratree) [inline api/src/main/java/cz/fi/muni/pa165/facade/HunterFacade.java:70]: same

File: persistence/pom.xml [hunk 8]
```diff
@@ -19,6 +19,11 @@
 
     <dependencies>
         <dependency>
+            <groupId>cz.fi.muni.pa165</groupId>
```
Comment (atiratree) [inline persistence/pom.xml]: this dependency is not needed

**Comment by atiratree** [comment]
Validation of entities, DataAccessExceptions missing

File: service/src/main/java/cz/fi/muni/pa165/service/HunterServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+    @Override
+    public boolean authenticate(Hunter hunter, String password) throws InvalidKeySpecException,
+            NoSuchAlgorithmException {
+
+        return verifyPassword(password, hunter.getPasswordHash());
+    }
+
+    @Override
+    public boolean isAdmin(Hunter hunter) {
```
Comment (atiratree) [inline service/src/main/java/cz/fi/muni/pa165/service/HunterServiceImpl.java]: unnecesary, please remove

File: service/src/main/java/cz/fi/muni/pa165/service/HunterServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        // Generate a random salt
+        SecureRandom random = new SecureRandom();
+        byte[] salt = new byte[SALT_BYTE_SIZE];
+        random.nextBytes(salt);
+
+        // Hash the password
+        byte[] hash = pbkdf2(unencryptedPassword.toCharArray(), salt, PBKDF2_ITERATIONS, HASH_BYTE_SIZE);
+        int hashSize = hash.length;
+
+        return PBKDF2_ITERATIONS + ":" + hashSize + ":" + toBase64(salt) + ":" + toBase64(hash);
```
Comment (atiratree) [inline service/src/main/java/cz/fi/muni/pa165/service/HunterServiceImpl.java]: is hashSize really needed? test sufficently

**Comment by atiratree** [comment]
Validation exceptions

File: service/src/main/java/cz/fi/muni/pa165/service/facade/HunterFacadeImpl.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public void delete(HunterDTO hunter) {
+        hunterService.delete(beanMapperService.mapTo(hunter,Hunter.class));
+    }
+
+    @Override
+    public HunterDTO findById(Long id) {
+        Hunter hunter = hunterService.findById(id);
+        return (hunter == null) ? null : beanMapperService.mapTo(hunter,HunterDTO.class);
```
Comment (atiratree) [inline service/src/main/java/cz/fi/muni/pa165/service/facade/HunterFacadeImpl.java:58]: test if mapper if mapper accepts null

File: service/src/main/java/cz/fi/muni/pa165/service/ForestServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+import javax.inject.Inject;
+import java.util.List;
+
+/**
+ * Created by Erik Macej 433744 , on 20.11.16.
+ *
+ * @author Erik Macej 433744
+ */
+@Service
+public class ForestServiceImpl implements ForestService {
```
Comment (atiratree) [inline service/src/main/java/cz/fi/muni/pa165/service/ForestServiceImpl.java:16]: DataAccessExcpetions everywhere

File: service/src/main/java/cz/fi/muni/pa165/service/HunterServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+        hunter.setPasswordHash(createHash(unencryptedPassword));
+        hunterDao.create(hunter);
+    }
+
+    @Override
+    public boolean authenticate(Hunter hunter, String password) throws InvalidKeySpecException,
+            NoSuchAlgorithmException {
+
+        return verifyPassword(password, hunter.getPasswordHash());
```
Comment (atiratree) [inline service/src/main/java/cz/fi/muni/pa165/service/HunterServiceImpl.java:40]: check for null

File: service/src/main/java/cz/fi/muni/pa165/service/HunterServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    @Override
+    public boolean isAdmin(Hunter hunter) {
+        return findById(hunter.getId()).getType().equals(Role.ADMIN);
+    }
+
+    @Override
+    public Hunter update(Hunter hunter) { return hunterDao.update(hunter); }
+
+    @Override
+    public void changePassword(Hunter hunter,String oldUnencryptedPassword, String newUnencryptedPassword)
```
Comment (atiratree) [inline service/src/main/java/cz/fi/muni/pa165/service/HunterServiceImpl.java:47]: check for null

File: service/src/test/java/cz/fi/muni/pa165/service/BeanMapperServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+    private VisitDTO visitDto2;
+    private HunterDTO hunterDto1;
+    private HunterDTO hunterDto2;
+    private MushroomDTO mushroomDto;
+    private MushroomCountDTO mushroomCountDto;
+    private ForestDTO forestDto;
+
+    @BeforeClass
+    public void prepareData(){
+        createVisits();
```
Comment (atiratree) [inline service/src/test/java/cz/fi/muni/pa165/service/BeanMapperServiceTest.java:42]: mapper: maybe only testing of few entities should be enough
also check nulls and special cases

**Pull request merged**